### ISSUE TITLE
`Doc.to_bytes` causing thinc AssertionError

### ISSUE THREAD

**Issue by notnami** [explosion/spaCy#3456]
## How to reproduce the behaviour

The below function works in Spark with Spacy 2.1.0, but crashes with Spacy 2.1.1. Rolling back to 2.1.0 gets rid of the error. 

```
import pandas as pd
from pyspark.sql.functions import pandas_udf
from pyspark.sql.types import BinaryType
import spacy

nlp = None

def get_spacy(spacy_model='en', spacy_disable=('parser',)):
    """
    Args:
        spacy_model (str, optional, defaults to 'en'): Name of Spacy model.
            This argument is used only if the Spacy model
            hasn't yet been initialised globally.
        spacy_disable (iterable of str, optional, default to ('parser',):
            Components of the Spacy pipeline that should be disabled.
            When unneeded components are disabled, Spacy can process documents faster.

    Returns:
        Language: Spacy object

    """

    global nlp
    if nlp is None:
        nlp = spacy.load(spacy_model, disable=spacy_disable)
    return nlp


@pandas_udf(returnType=BinaryType())
def get_docs_bytes(series: pd.Series) -> pd.Series:
    nlp = get_spacy()   
    preprocessed = series.apply(preprocess_text)  # this just does some string preprocessing
    docs = nlp.pipe(preprocessed, n_threads=cpu_count(), batch_size=10000)
    docs_bytes = pd.Series([doc.to_bytes() for doc in docs])  # this is where the error happens
    return docs_bytes
```

Here is the error:

```
"/mnt/yarn/usercache/hadoop/appcache/application_1553100215590_0001/container_1553100215590_0001_01_000001/nlp_insights.zip/nlp_insights/core.py", line 116, in get_docs_bytes
  File "/mnt/yarn/usercache/hadoop/appcache/application_1553100215590_0001/container_1553100215590_0001_01_000001/nlp_insights.zip/nlp_insights/core.py", line 116, in <listcomp>
  File "/usr/local/lib64/python3.6/site-packages/spacy/language.py", line 711, in pipe
    for doc in docs:
  File "/usr/local/lib64/python3.6/site-packages/spacy/language.py", line 898, in _pipe
    for doc in docs:
  File "/usr/local/lib64/python3.6/site-packages/spacy/language.py", line 898, in _pipe
    for doc in docs:
  File "nn_parser.pyx", line 221, in pipe
  File "/usr/local/lib64/python3.6/site-packages/spacy/util.py", line 457, in minibatch
    batch = list(itertools.islice(items, int(batch_size)))
  File "pipes.pyx", line 380, in pipe
  File "pipes.pyx", line 392, in spacy.pipeline.pipes.Tagger.predict
  File "/usr/local/lib64/python3.6/site-packages/thinc/neural/_classes/model.py", line 165, in __call__
    return self.predict(x)
  File "/usr/local/lib64/python3.6/site-packages/thinc/neural/_classes/feed_forward.py", line 40, in predict
    X = layer(X)
  File "/usr/local/lib64/python3.6/site-packages/thinc/neural/_classes/model.py", line 165, in __call__
    return self.predict(x)
  File "/usr/local/lib64/python3.6/site-packages/thinc/api.py", line 280, in predict
    return layer.ops.unflatten(X, lengths, pad=pad)
  File "ops.pyx", line 138, in thinc.neural.ops.Ops.unflatten
AssertionError
```

## Your Environment
* Python Version Used: 3.6.5
* spaCy Version Used: 2.1.1
* Environment Information: PySpark 2.4.0 in Amazon EMR 5.21.0

Please let me know if you need more information.

**Comment by mwakaba2** [comment]
Could this be the cause?
#3431 
https://github.com/explosion/spaCy/pull/3440/files

**Comment by honnibal** [comment]
@mwakaba2 I think you're right, thanks for the help. If you have time to make a PR that would be great.

**Comment by mwakaba2** [comment]
@honnibla thanks, I'll give it a try. 👍

**Comment by honnibal** [comment]
@mwakaba2 I'm digging a bit deeper into this and I'm no longer sure that's the problem.

@notnami I think the `doc.to_bytes()` is a red herring here, because what's actually failing is the prediction in `nlp.pipe()`. It's a generator, so it's only getting called as the data streams through the loop below in your code. 

Obviously a naked `assert` isn't great. Here's the function and the error we're hitting:

```
    def unflatten(self, X, lengths, pad=0):
        unflat = []
        pad = int(pad)
        for length in lengths:
            length = int(length)
            if pad >= 1 and length != 0:
                X = X[pad:]
            unflat.append(X[:length])
            X = X[length:]
        if pad >= 1 and length != 0:
            X = X[pad:]
        assert len(X) == 0
        assert len(unflat) == len(lengths)
        return unflat
```

The function performs a common data transformation for Thinc: it converts from a flattened representation, where we concatenate the arrays together and we keep the lengths separately, to a normal list of arrays.

We're failing at `assert len(X) == 0` check, which makes sure we've actually consumed all the data from `X`. This suggests we're somehow ending up with a lengths array that's wrong for the data array. It's not obvious to me how that could happen. Perhaps an empty string? I thought we tested for that, but maybe not in the tagger. It would be very helpful to know if it still fails if you just call `nlp(text)` instead of piping, and if so, which doc it fails on.

**Comment by honnibal** [comment]
Btw, you might find the code in https://github.com/explosion/spaCy/blob/master/spacy/tokens/_serialize.py useful. We haven't finalised this yet, because backwards incompatibilities in serialization really suck once you've parsed a lot of text. But I've been using it a bit and it works quite well.

**Comment by jeffoneill** [comment]
Here is an easy way to recreate this error:

    nlp = spacy.load('en_core_web_sm')
    list(nlp.pipe(['hi', '']))

* **spaCy version:** 2.1.2
* **Platform:** Darwin-17.7.0-x86_64-i386-64bit
* **Python version:** 3.6.5

**Comment by notnami** [comment]
It might in fact be an empty string because when we drop all empty strings, the error no longer appears.

Anton

**Comment by mwakaba2** [comment]
Hi @honnibal, yes you're right. I'm able to reproduce the error with jeffoneill's example.
I can close this ticket and create a new one for the thinc/empty string issue?

**Comment by mwakaba2** [comment]
@honnibal So here's an update.  I think the issue is in this thinc function `ops.pyx in thinc.neural.ops.Ops.unflatten()`.

 I looked at the `X` and `lengths` parameters when I pass in the following to `nlp.pipe`
```python
    sentences = ["hi", ""]
    docs = nlp.pipe(sentences, batch_size=2)
    docs_list = list(docs)
```

`layer.ops.unflatten(X, lengths, pad=pad)` is getting the following inputs
```bash
X.shape --> (9, 96)
lengths --> [1, 0]
pad --> 4
```
The error happens on the following line
```python
print(len(X), len(unflat), len(lengths)) # This outputs 4, 2, 2
assert len(X) == 0
```

I started working on a test in the thinc repository. Hopefully I can get a PR ready today.
